timely/src/progress/broadcast.rs

```diff
@@ -58,34 +59,26 @@ impl<T:Timestamp+Send> Progcaster<T> {
         changes.compact();
         if !changes.is_empty() {
 
-            // This logging is relatively more expensive than other logging, as we
-            // have formatting and string allocations on the main path. We do have
-            // local type information about the timestamp, and we could log *that*
-            // explicitly, but the consumer would have to know what to look for and
-            // interpret appropriately. That's a big ask, so let's start with this,
-            // and as folks need more performant logging think about allowing users
-            // to select the more efficient variant.
-            self.logging.as_ref().map(|l| {
+            self.progress_logging.as_ref().map(|l| {
 
                 // Pre-allocate enough space; we transfer ownership, so there is not
                 // an apportunity to re-use allocations (w/o changing the logging
                 // interface to accept references).
-                let mut messages = Vec::with_capacity(changes.len());
-                let mut internal = Vec::with_capacity(changes.len());
+                let mut messages = Box::new(Vec::with_capacity(changes.len()));
+                let mut internal = Box::new(Vec::with_capacity(changes.len()));
```

Comment:
I'm super confused that this works. I would have thought that as soon as this is boxed up you lose access to the `push` methods. Or is it because there is an auto-conversion from `Box<T>` to `Box<dyn Trait>` for any `T: Trait`? No change needed, just surprising to me!

Reply:
I believe `push` works because of `https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-Deref` (`Deref` impl for `Box`). And yes, you can upcast a `Box<T>` to a `Box<dyn Trait>` for the same reason you can upcast a `&T` to a `&dyn Trait` for `T: Trait`.

Reply:
To clarify, the type of `messages` and `internal` is still the concrete, non-`dyn` vector type. The upcast happens on assignment to the target struct below.